pkg/registry/core/service/strategy.go

```diff
@@ -19,54 +19,139 @@ package service
+			NameGenerator: names.SimpleNameGenerator,
+			ipFamilies:    []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol},
+		}
+	case hasSecondary && serviceIPFamily == api.IPv6Protocol:
+		strategy = svcStrategy{
+			ObjectTyper:   legacyscheme.Scheme,
+			NameGenerator: names.SimpleNameGenerator,
+			ipFamilies:    []api.IPFamily{api.IPv6Protocol, api.IPv4Protocol},
+		}
+	case serviceIPFamily == api.IPv6Protocol:
+		strategy = svcStrategy{
+			ObjectTyper:   legacyscheme.Scheme,
+			NameGenerator: names.SimpleNameGenerator,
+			ipFamilies:    []api.IPFamily{api.IPv6Protocol},
+		}
+	default:
+		strategy = svcStrategy{
+			ObjectTyper:   legacyscheme.Scheme,
+			NameGenerator: names.SimpleNameGenerator,
+			ipFamilies:    []api.IPFamily{api.IPv4Protocol},
+		}
+	}
+	return strategy, serviceIPFamily
+}
 
 // NamespaceScoped is true for services.
 func (svcStrategy) NamespaceScoped() bool {
 	return true
 }
 
 // PrepareForCreate clears fields that are not allowed to be set by end users on creation.
-func (svcStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+func (strategy svcStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
 	service := obj.(*api.Service)
 	service.Status = api.ServiceStatus{}
 
+	// perform contextual defaulting to the configured family, if no family has been defaulted
+	if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		serviceHasCustomEndpoints := service.Spec.ClusterIP == api.ClusterIPNone && len(service.Spec.Selector) == 0
+		ipFamilyForbidden := service.Spec.Type == api.ServiceTypeExternalName
+		ipFamilyNotDefaulted := ipFamilyForbidden || serviceHasCustomEndpoints
+		if service.Spec.IPFamily == nil {
+			if ipFamilyNotDefaulted {
+			} else {
+				family := strategy.ipFamilies[0]
+				service.Spec.IPFamily = &family
+			}
+		} else {
+			if ipFamilyForbidden {
+				service.Spec.IPFamily = nil
+			}
+		}
+	}
+
 	dropServiceDisabledFields(service, nil)
 }
 
 // PrepareForUpdate clears fields that are not allowed to be set by end users on update.
-func (svcStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+func (strategy svcStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
 	newService := obj.(*api.Service)
 	oldService := old.(*api.Service)
 	newService.Status = oldService.Status
 
+	// perform contextual defaulting to the configured family, if no family has been defaulted
+	if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		serviceHasCustomEndpoints := newService.Spec.ClusterIP == api.ClusterIPNone && len(newService.Spec.Selector) == 0
+		ipFamilyForbidden := newService.Spec.Type == api.ServiceTypeExternalName
+		ipFamilyNotDefaulted := ipFamilyForbidden || serviceHasCustomEndpoints
+		if newService.Spec.IPFamily == nil {
+			if ipFamilyNotDefaulted {
+			} else {
+				if oldService.Spec.IPFamily != nil {
+					newService.Spec.IPFamily = oldService.Spec.IPFamily
```

Comment:
this whole block only runs if the feature is on... which means a client that drops the ipfamily field will fail if it tries to update a `type=ExternalName,clusterIP="1.2.3.4",ipFamily=ipv4` service to `type=ExternalName,clusterIP=""`? see matching comment in validateIPFamily

Reply:
Yeah, that was a bug before.